unit_test/degrib2c.c

```diff
@@ -0,0 +1,168 @@
+
+      g2int  msk1=32000;
+
+      //start();
+      unpack=1;
+      expand=0;
+      
+// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+//  GET ARGUMENTS
+      if (argc != 2) {
+        printf("degrib2c:  Incorrect usage\n");
+        printf("Usage: degrib2c grib2file\n");
+        return 2;
+      }
+
+      ifl1=fopen(argv[1],"r");
+      if ( ifl1 == 0 ) {
+         perror("cannot open file:");
+         return 3;
+      }
+
+      itot=0;
+      icount=0;
+      iseek=0;
+
+      for (;;) {
+         seekgb(ifl1,iseek,msk1,&lskip,&lgrib);
+         if (lgrib == 0) break;    // end loop at EOF or problem
+         if (lgrib > currlen) {
+            if (cgrib!=0) free(cgrib);
+            cgrib=(unsigned char *)malloc(lgrib);
+            currlen=lgrib;
+         }
+         //call baread(ifl1,lskip,lgrib,lengrib,cgrib)
+         ret=fseek(ifl1,lskip,SEEK_SET);
+         lengrib=fread(cgrib,sizeof(unsigned char),lgrib,ifl1);
+         if (lgrib != lengrib) {
+            printf(" degrib2c: IO Error.\n");
+            return 9;
+         }
+         iseek=lskip+lgrib;
+         icount++;
+
+         printf("\n GRIB MESSAGE  %ld starts at %ld \n\n",icount,lskip+1);
+
+// Unpack GRIB2 field
+         ierr=g2_info(cgrib,listsec0,listsec1,&numfields,&numlocal);
+         if (ierr != 0) {
+           printf(" ERROR querying GRIB2 message = %d\n",ierr);
+           return 10;
+         }
+         itot=itot+numfields;
+         printf("  SECTION 0:  ");
+         for (j=0;j<3;j++) printf("%ld ",listsec0[j]);
+         printf("\n");
+         printf("  SECTION 1:  ");
+         for (j=0;j<13;j++) printf("%ld ",listsec1[j]);
+         printf("\n");
+         printf("  Contains  %ld  Local Sections  and  %ld  data fields.\n",
+                numlocal,numfields);
+
+         for (n=0;n<numfields;n++) {
+           ierr=g2_getfld(cgrib,n+1,unpack,expand,&gfld);
+           if (ierr != 0) {
+             printf(" ERROR extracting field = %d\n",ierr);
+             continue;
+           }
+
+           printf("\n  FIELD  %ld\n",n+1);
+           if (n==0) {
+             printf("  SECTION 0:  %ld %ld\n",gfld->discipline,gfld->version);
+             printf("  SECTION 1:  ");
+             for (j=0;j<gfld->idsectlen;j++) printf("%ld ",gfld->idsect[j]);
+             printf("\n");
+           }
+           if ( gfld->local!=0 && gfld->locallen>0 ) {
+              printf("  SECTION 2: %d bytes\n",gfld->locallen);
+              /*for (j=0;j<gfld->locallen;j++) printf("%c",gfld->local[j]);
+              printf("\n");*/
+           }
+           printf("  SECTION 3:  %ld %ld %ld %ld %ld\n",
+                    gfld->griddef,gfld->ngrdpts,gfld->numoct_opt,
+                    gfld->interp_opt,gfld->igdtnum);
+           printf("  GRID TEMPLATE 3. %ld :  ",gfld->igdtnum);
+           for (j=0;j<gfld->igdtlen;j++) printf("%ld ",gfld->igdtmpl[j]);
+           printf("\n");
+           if ( gfld->num_opt == 0 ) 
```

Comment:
For each test file, you should know what optional sections are present, and make sure they are handled correctly.